Build: Fix Redis/Memcached integration tests

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Cache tests didn't use to run: they were ignored in OSS and in enterprise the integration tests used to re-run instead of them

**Why do we need this feature?**

This fix:
- removes build tags from OSS cache tests
- skips these tests when `-test.short` flag is set and when the expected environmental variables are not set (`REDIS_URL` for the redis and `MEMCACHED_HOSTS` for the memcached ones)
- updates release CI pipeline to run cache  redis/memcached instead of integration tests
- adds redis/memcached tests in the other pipelines (pr, main)

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #60587

**Special notes for your reviewer**:
for testing locally:
```sh
make devenv sources=redis,memcached
make test-go-integration-redis
make test-go-integration-memcached
```

the redis tests used to fail therefore ~I had to make [these changes](https://github.com/grafana/grafana/pull/64298/commits/5350751673367ea0a808538828c50e7118561a2e) but I'm not at ll sure if the cache is supposed to behave this way.~ I have reverted this change and I have modified the tests instead

papagian committed:
Build: Fix integration cache tests

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-60587

papagian committed:
fixup

papagian committed:
fix fix

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-60587

papagian committed:
Fixes so that redis tests do not fail

papagian commented:
it seems there is inconsistency on the expected redis URL in [OSS](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#redis) and [enterprise](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/enterprise-configuration/#url-3)
so either the tests should accept both or adopt the same (breaking change)

papagian committed:
Revert "Fixes so that redis tests do not fail"

This reverts commit 5350751673367ea0a808538828c50e7118561a2e.

papagian committed:
Make redis tests not to fail

papagian committed:
Allow REDIS_URL with scheme

papagian committed:
Fix redis URL parsing

papagian committed:
Fix starlack failure

papagian committed:
Revert test change

`testFrameCache` is used by both `TestMemoryFrameCache`
and `TestIntegrationRedisCacheStorage` but the later fails
because `result["schema"] != stringSchema` is false

papagian committed:
More starlark fixing

papagian commented on pkg/infra/remotecache/remotecache_test.go:
```diff
@@ -95,7 +95,8 @@ func canPutGetAndDeleteCachedObjects(t *testing.T, client CacheStorage) {
 
 	_, err = client.Get(context.Background(), "key1")
-	assert.Equal(t, err, ErrCacheItemNotFound)
+	// redis client returns redis.Nil error when key does not exist.
+	assert.Error(t, err)
```
not sure if the redis implementation should change instead and return `ErrCacheItemNotFound` in that case

bergquist replied:
Sounds like a good idea but perhaps in another PR.

zerok commented on scripts/drone/steps/lib.star:
```diff
@@ -1143,7 +1143,7 @@ def redis_integration_tests_step():
             "dockerize -wait tcp://redis:6379/0 -timeout 120s",
             "go clean -testcache",
-            "go list './pkg/...' | xargs -I {} sh -c 'go test -run Integration -covermode=atomic -timeout=5m {}'",
+            "go test -run IntegrationRedis -covermode=atomic -timeout=30m ./pkg/...",
```
Do you really need a timeout of 30 minutes here?

zerok commented on scripts/drone/services/services.star:
```diff
@@ -36,21 +36,20 @@ def integration_test_services(edition):
-                },
-            ],
-        )
+    services.extend(
```
Please add them directly to the service list above if this is no longer conditional 🙂

zerok commented on scripts/drone/events/release.star:
```diff
@@ -266,7 +266,7 @@ def enterprise_pipelines(ver_mode = ver_mode, trigger = release_trigger):
     environment = {"EDITION": "enterprise"}
 
-    services = integration_test_services(edition = "enterprise")
+    services = integration_test_services()
```
Since now *all* services are launched, does it also make sense to run the redis/memcached integration tests for non-enterprise builds (line 338)?

papagian replied:
not sure I understand: do you mean since in enterprise builds will run also the OSS ones we don't need to run them for non-enterprise builds?

kminehart replied:
I think Horst is suggesting that redis/memcached tests might not be running for non-enterprise builds. For context, when I added them, enterprise was the only one that had redis/memcached integration tests.

papagian replied:
thank for the clarification: I have addeed them also in the OSS builds

zerok reviewed (changes requested):
Thank you so much for working on this ❤️

DanCech commented on .drone.yml:
```diff
@@ -797,6 +797,12 @@ services:
   - name: mysql
     path: /var/lib/mysql
+- environment: {}
+  image: redis:6.2.1-alpine
```
this is a 2 year old version, should we be using 6.2.11 (latest 6.2.x) or 7.0.10 (current stable) instead?

papagian replied:
it seems that in cloud we use [another image](https://github.com/grafana/deployment_tools/blob/28f2a0d86338e8150a57814fdfa0e1c53461db8a/ksonnet/lib/hosted-grafana/images.libsonnet#L21)
it would make sense to use the same, @grafana/grafana-delivery WDYT?

kminehart replied:
i think it makes sense to use the same.

papagian replied:
I have ended up using the `redis:6.2.11-alpine` for now because the image used by cloud [requires extra setup for setting a password](https://drone.grafana.net/grafana/grafana/109102/8/4) so it requires extra work for modifying the `REDIS_URL` for test to include it

kminehart commented:
> it seems there is inconsistency on the expected redis URL in [OSS](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#redis) and [enterprise](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/enterprise-configuration/#url-3) so either the tests should accept both or adopt the same (breaking change)

You could allow for both in Enterprise and make this a non-breaking change.

In Enterprise, `redis.ParseURL` https://pkg.go.dev/github.com/go-redis/redis#ParseURL is used to parse the URL. I imagine, if that returns an error, then we could attempt to parse the URL the same way as OSS does.

papagian commented:
> > it seems there is inconsistency on the expected redis URL in [OSS](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#redis) and [enterprise](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/enterprise-configuration/#url-3) so either the tests should accept both or adopt the same (breaking change)
> 
> You could allow for both in Enterprise and make this a non-breaking change.
> 
> In Enterprise, `redis.ParseURL` https://pkg.go.dev/github.com/go-redis/redis#ParseURL is used to parse the URL. I imagine, if that returns an error, then we could attempt to parse the URL the same way as OSS does.
> 
> I don't know for sure but I have a feeling that `ParseURL` will also handle connection strings. But I'm just guessing

yes, that's exactly what I have ended up [doing](https://github.com/grafana/grafana/pull/64298/files#diff-667c888dc0457bb7078b721c9798117ccc9deadd9b8e462ebc0209bd14774444)

kminehart commented:
ah I didn't see that! awesome, don't mind me then :joy:

papagian committed:
Use same redis image as cloud

papagian committed:
Reduce cache integration tests timeout to 5m

papagian committed:
fixup

papagian committed:
reduce timeout even more

papagian committed:
Apply suggestion from code review

papagian committed:
Run redis/memcached integration tests in OSS pipelines

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-60587

papagian committed:
Update drone after merging with main

papagian committed:
Restore accidentally committed line

papagian committed:
Change redis image

Revert using bitnami/redis image for bow because it requires extra setup
for setting a password

DanCech reviewed (approved):
Nice!
